### PR TITLE
gmp: modernize

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **gmp/all**

for https://github.com/conan-io/hooks/pull/320

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by Croydon — DISMISSED** [review]

**Review by prince-chrismc — DISMISSED** [review]

File: recipes/gmp/all/conanfile.py [hunk 1]
```diff
@@ -49,19 +67,18 @@ def package_id(self):
...
+            self.build_requires("msys2/cci.latest")
         if self.settings.compiler == "Visual Studio":
             self.build_requires("yasm/1.3.0")
-            self.build_requires("automake/1.16.3")
+            self.build_requires("automake/1.16.4")
 
     def source(self):
         tools.get(**self.conan_data["sources"][self.version], destination=self._source_subfolder, strip_root=True)
 
+    @functools.lru_cache(1)
```
Comment (prince-chrismc) [inline recipes/gmp/all/conanfile.py:79]: Is this needed everywhere?

File: recipes/gmp/all/conanfile.py [hunk 2]
```diff
@@ -83,15 +100,15 @@ def _configure_autotools(self):
                 "gmp_cv_asm_label_suffix=:",
                 "lt_cv_sys_global_symbol_pipe=cat",  # added to get further in shared MSVC build, but it gets stuck later
             ])
-            self._autotools.flags.append("-FS")
-            # self._autotools.cxx_flags.append("-EHsc")
-        self._autotools.configure(args=configure_args, configure_dir=self._source_subfolder)
-        return self._autotools
+            autotools.flags.append("-FS")
```
Comment (prince-chrismc) [inline recipes/gmp/all/conanfile.py:103]:
```suggestion
            if tools.Version(self.settings.compiler.version) >= 12:
                autotools.flags.append("-FS")
```

File: recipes/gmp/all/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/gmp/all/conanfile.py:79]: Not needed, only if you want to be fancy. 😅✌️

**Review by prince-chrismc — APPROVED** [review]

File: recipes/gmp/all/conanfile.py [hunk 3]
```diff
@@ -11,14 +12,27 @@ class GmpConan(ConanFile):
...
+        "fPIC": True,
+        "disable_assembly": True,
+        "enable_fat": False,
+        "run_checks": False,
+        "enable_cxx": True,
+    }
+
     exports_sources = "patches/*"
 
     _autotools = None
```
Comment (uilianries) [inline recipes/gmp/all/conanfile.py:38]:
```suggestion
```

lru_cache

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**